### ISSUE TITLE
parse JSONLines format

### ISSUE THREAD

**Issue by saulpw** [kellyjonbrazil/jello#50]
Hi, is it possible to parse JSONLines format with jello into a top-level array?

**Comment by kellyjonbrazil** [comment]
Hi there! This should already be possible today. Jello accepts JSON and JSON lines as input. Let me know if you run into any issues.

**Comment by saulpw** [comment]
Hm, I'm using v1.5.4 and I get:

```
jello:  JSON Load Exception:  JSONDecodeError
        Extra data: line 2 column 1 (char 1048)
```

Is there an option I have to specify?  Or is it only available on a later version?

**Comment by kellyjonbrazil** [comment]
Looks like it's not valid JSON Lines. Could you copy/paste the first few lines?

**Comment by saulpw** [comment]
Okay, so it turns out that this JSONL has a few blank lines in it, which seems to be what's tripping it up.  Ideally `jello` would ignore blank lines, which happen sometimes in the best of files :)

But, perhaps more importantly, these lines are thousands of lines into the file, and `jello` doesn't output anything for the lines before it.  (Which is why I got tripped up by the error message; it's not actually line 2 in the file).  I guess `jello` tries to parse everything before it will do anything with the result?  I was hoping to use it to process rows from a .jsonl file in a streaming manner.

**Comment by kellyjonbrazil** [comment]
Ah cool - as it happens I was just looking at a more `permissive` style of slurping the JSON objects. I'll see if that helps with these types of things.

Unfortunately `jello` is not a good option for super large datasets as it is not streaming and I'm not really sure of a great way to allow it and still have simple/standard python syntax when querying. I'm open to ideas.

I've looked into streaming JSON libraries, but then the api in now you interact with the data changes considerably, so you can't use normal list/dict type of patterns. Maybe that's ok, though, with a streaming option?

**Comment by saulpw** [comment]
Could you treat a JSONL file specifically as a series of independent objects, instead of an array?  Then the Python would only refer to the current row, and each row could be processed as it is parsed.  The only functionality you'd lose, I think, is the ability to easily select the Nth row of the file; but you can already do that pretty easily with head/tail since it's line-based.

It would also be really great to be able to do the same kind of streaming processing on JSON arrays, but I know it's a lot harder to use existing libraries to process them in a reasonable way.  Sometimes though I have a JSON file that's just a JSONL file wrapped with `[`/`]` and with a trailing comma at the end of each line, and it's annoying to have to parse the whole blob in one go, so I actually use text-processing commands to turn it into proper JSONL.  There has to be a better way...

**Comment by kellyjonbrazil** [comment]
Yeah, that could be an option - maybe you just lose some python functionality when  streaming. It might be good for some use cases.

I was actually testing this out a while back with my own implemented streaming method. Actually quite a few different methods to see if I could get decent performance. I think the best I could do was around half as fast as `jq` so I kinda just left it. I was able to get it a little faster with another JSON library like `orjson` (about 2/3's the speed of `jq`) but it would get even slower with nested JSON objects. And even then I wasn't quite sure how to use it, other than like you said, as discrete JSON objects one by one.

The performance penalty might not be so bad if it helps with memory consumption. Also, like I mentioned before, I came across another method to pull in a stream of JSON objects I want to try out. It might possibly be faster than my pretty naive implementations.

**Comment by rjgladish** [comment]
A python library known as ijson https://pypi.org/project/ijson/ is more
permissive than typical JSON parsers, including truncated json files. To do
this, ijson invokes the libyajl.so library.

**Comment by kellyjonbrazil** [comment]
I also like the idea of `de-slurping` (is that a word?) :). So the idea is to take a JSON array of objects and just ignore the array delimiters so you can stream the input. Interesting!

**Comment by kellyjonbrazil** [comment]
> A python library known as ijson https://pypi.org/project/ijson/ is more permissive than typical JSON parsers, including truncated json files. To do this, ijson invokes the libyajl.so library.

Cool, thanks - I'll look into that!

**Comment by kellyjonbrazil** [comment]
I wonder if it would make sense to abstract the python library that `jello` uses. That way by default you just get the standard json library, but with some sort of flag you can specify another one that is installed on the system so you can use that library's syntax to query?

**Comment by saulpw** [comment]
> I also like the idea of de-slurping (is that a word?) :). So the idea is to take a JSON array of objects and just ignore the array delimiters so you can stream the input. Interesting!

You also have to ignore the trailing commas, and they do have to each be on one line, which is sometimes not the case.  But yeah when it works it's a neat hack :)

I dream of a performant streaming JSON parser that can handle these oddities and more and report them as errors out-of-band, and then continue processing.  (I also want this same thing for XML, with an identical API.)   Someday..

**Comment by kellyjonbrazil** [comment]
Looks like the `ijson` library might handle a lot of this. I think I might be able to create a `jello -S` (for streaming) option that uses the `ijson` library to parse `STDIN` and return `_` as a generator/iterator of JSON objects - whether it's an array or a top-level of JSON objects.

**Comment by kellyjonbrazil** [comment]
I have a fix in `dev` that will allow the loading of JSON Lines data if there happens to be extra blank lines and other whitespace between items. This will go into `jello` v1.5.5.

I'll open a new issue for streaming support.

**Comment by kellyjonbrazil** [comment]
fixed in v1.5.5.
